source/fuzz/fuzzer_pass_replace_irrelevant_ids.cpp

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/fuzzer_pass_replace_irrelevant_ids.h"
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/id_use_descriptor.h"
+#include "source/fuzz/transformation_replace_irrelevant_id.h"
+
+namespace spvtools {
+namespace fuzz {
+
+// A fuzzer pass that, for every use of an id that has been recorded as
+// irrelevant, randomly decides whether to replace it with another id of the
+// same type.
+FuzzerPassReplaceIrrelevantIds::FuzzerPassReplaceIrrelevantIds(
+    opt::IRContext* ir_context, TransformationContext* transformation_context,
+    FuzzerContext* fuzzer_context,
+    protobufs::TransformationSequence* transformations)
+    : FuzzerPass(ir_context, transformation_context, fuzzer_context,
+                 transformations) {}
+
+FuzzerPassReplaceIrrelevantIds::~FuzzerPassReplaceIrrelevantIds() = default;
+
+void FuzzerPassReplaceIrrelevantIds::Apply() {
+  // Keep track of the irrelevant ids.
+  std::vector<uint32_t> irrelevant_ids;
+
+  // Keep a map from the type ids of irrelevant ids to all the ids with that
+  // type.
+  std::unordered_map<uint32_t, std::vector<uint32_t>> types_to_ids;
+
+  for (auto id :
+       GetTransformationContext()->GetFactManager()->GetIrrelevantIds()) {
+    // Check that the id still exists in the module.
+    auto declaration = GetIRContext()->get_def_use_mgr()->GetDef(id);
+    if (!declaration) {
+      continue;
+    }
+
+    irrelevant_ids.push_back(id);
+
+    // If the type of this id has not been seen before, add a mapping from this
+    // type id to an empty list in |types_to_ids|. The list will be filled later
+    // on.
+    if (types_to_ids.count(declaration->type_id()) == 0) {
+      types_to_ids[declaration->type_id()] = {};
+    }
+  }
+
+  // If no irrelevant ids were found, return.
+  if (irrelevant_ids.empty()) {
+    return;
+  }
+
+  // Loop through all the ids in the module and fill all the empty lists in
+  // |types_to_ids|.
+  for (const auto& pair : GetIRContext()->get_def_use_mgr()->id_to_defs()) {
+    uint32_t type_id = pair.second->type_id();
+    if (types_to_ids.count(type_id)) {
```

Comment:
I guess you are relying on the fact that if there is no type, type_id() will return 0 and 0 will not be a key for your map.  I suppose that's fine.  I would probably write:

if(type_id && types_to_ids.count(type_id))

personally.

Reply:
I agree that this makes it clearer